invalid character on typedef name generated in C: typedef array Array_main__Node<>;

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, "Fedora release 33 (Thirty Three)"
Processor: 4 cpus, 64bit, little endian, Intel(R) Celeron(R) CPU  N3160  @ 1.60GHz
CC version: cc (GCC) 10.3.1 20210422 (Red Hat 10.3.1-1)

getwd: /home/notecomm/pv/data_structures
vmodules: /home/notecomm/.vmodules
vroot: /home/notecomm/v
vexe: /home/notecomm/v/v
vexe mtime: 2021-07-06 10:45:45
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.2 1e6230a

Git version: git version 2.31.1
Git vroot status: weekly.2021.27-6-g1e6230a4
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 7543de81
```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg t.v`
```v
struct Node<T> {
mut:
  data T
  next &Node<T> = 0
}

struct SinglyLinkedList<T> {
  mut: first_node &Node<T> = 0
}

fn init_singlylinkedlist<T>(nodes []Node<T>) SinglyLinkedList<T> {
  mut current_node := &nodes[0]

  for i in 0..nodes.len - 1 {
    current_node = &nodes[i]
    current_node.next = &nodes[i + 1]
  }

  return SinglyLinkedList{&nodes[0]}
}

fn main() {
  sll := init_singlylinkedlist<int>([Node<int>{data: 1}, Node<int>{data: 2}, Node<int>{data: 798}])
  println(sll.first_node.next)
}
```

**What did you expect to see?**

&Node<int>{data: 2, next: &circular}


**What did you see instead?**
```
==================
/tmp/v/t.5189433308795789637.tmp.c:813: error: ';' expected (got "<")
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

If you were not working with C interop, this is a compiler bug, please report the bug using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```
```c
// /tmp/v/t.5189433308795789637.tmp.c:813
typedef array Array_main__Node<>;
```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -769,7 +769,11 @@ pub fn (mut g Gen) write_typedef_types() {
 		}
 		match typ.kind {
 			.array {
-				g.type_definitions.writeln('typedef array $typ.cname;')
+				info := typ.info as ast.Array
+				elem_sym := g.table.get_type_symbol(info.elem_type)
+				if elem_sym.kind != .placeholder {
+					g.type_definitions.writeln('typedef array $typ.cname;')
+				}
 			}
 			.array_fixed {
 				info := typ.info as ast.ArrayFixed

--- a/vlib/v/tests/generics_array_typedef_test.v
+++ b/vlib/v/tests/generics_array_typedef_test.v
@@ -1,0 +1,31 @@
+struct Node<T> {
+mut:
+	data T
+	next &Node<T> = 0
+}
+
+struct SinglyLinkedList<T> {
+mut:
+	first_node &Node<T> = 0
+}
+
+fn init_singlylinkedlist<T>(nodes []Node<T>) SinglyLinkedList<T> {
+	mut current_node := &nodes[0]
+
+	for i in 0 .. nodes.len - 1 {
+		current_node = &nodes[i]
+		current_node.next = &nodes[i + 1]
+	}
+
+	return SinglyLinkedList<T>{&nodes[0]}
+}
+
+fn test_generic_array_typedef() {
+	sll := init_singlylinkedlist<int>([Node<int>{ data: 1 }, Node<int>{
+		data: 2
+	}, Node<int>{
+		data: 798
+	}])
+	println(sll.first_node.next)
+	assert sll.first_node.next.data == 2
+}

